modules/core/src/test/java/org/apache/ignite/internal/processors/compute/ComputeGridMonitorMonitorTest.java

```diff
@@ -0,0 +1,431 @@
+
+        stopAllGrids();
+
+        IgniteEx crd = startGrids(2);
+
+        crd.cluster().state(ACTIVE);
+
+        awaitPartitionMapExchange();
+
+        CRD = crd;
+    }
+
+    /** {@inheritDoc} */
+    @Override protected void afterTestsStopped() throws Exception {
+        super.afterTestsStopped();
+
+        stopAllGrids();
+    }
+
+    /** {@inheritDoc} */
+    @Override protected void beforeTest() throws Exception {
+        super.beforeTest();
+
+        CRD.context().task().listenStatusUpdates(monitor = new ComputeGridMonitorImpl());
+    }
+
+    /** {@inheritDoc} */
+    @Override protected void afterTest() throws Exception {
+        super.afterTest();
+
+        CRD.context().task().stopListenStatusUpdates(monitor);
+    }
+
+    /** {@inheritDoc} */
+    @Override protected IgniteConfiguration getConfiguration(String igniteInstanceName) throws Exception {
+        return super.getConfiguration(igniteInstanceName)
+            .setFailureHandler(new StopNodeFailureHandler());
+    }
+
+    /**
+     * Checking get of diffs for the successful execution of the task.
+     */
+    @Test
+    public void simpleTest() {
+        ComputeTaskFuture<Void> taskFut = CRD.compute().executeAsync(new NoopComputeTask(), null);
+
+        taskFut.get(getTestTimeout());
+
+        assertTrue(monitor.statusSnapshots.isEmpty());
+
+        assertEquals(3, monitor.statusDiffs.size());
+
+        checkStartTask(monitor.statusDiffs.poll(), taskFut.getTaskSession());
+        checkChangeJobNodes(monitor.statusDiffs.poll(), taskFut.getTaskSession());
+        checkFinishTask(monitor.statusDiffs.poll(), taskFut.getTaskSession());
+    }
+
+    /**
+     * Checking get of diffs for the failed execution of the task.
+     */
+    @Test
+    public void failTaskTest() {
+        NoopComputeTask task = new NoopComputeTask() {
+            /** {@inheritDoc} */
+            @Override public Void reduce(List<ComputeJobResult> results) throws IgniteException {
+                throw new IgniteException("FAIL TASK");
+            }
+        };
+
+        ComputeTaskFuture<Void> taskFut = CRD.compute().executeAsync(task, null);
+
+        assertThrows(log, () -> taskFut.get(getTestTimeout()), IgniteException.class, null);
+
+        assertTrue(monitor.statusSnapshots.isEmpty());
+
+        assertEquals(3, monitor.statusDiffs.size());
+
+        checkStartTask(monitor.statusDiffs.poll(), taskFut.getTaskSession());
+        checkChangeJobNodes(monitor.statusDiffs.poll(), taskFut.getTaskSession());
+        checkFailTask(monitor.statusDiffs.poll(), taskFut.getTaskSession());
+    }
+
+    @Test
+    public void changeAttributesTest() throws Exception {
+        ComputeFullWithWaitTask task = new ComputeFullWithWaitTask(getTestTimeout());
+
+        ComputeTaskFuture<Void> taskFut = CRD.compute().executeAsync(task, null);
+
+        task.doneOnMapFut.get(getTestTimeout());
+
+        taskFut.getTaskSession().setAttribute("test", "test");
```

Comment:
Should there be a sleep after this line? How do we guarantee that attributes were actually set?

Reply:
Add **ComputeTaskSession#waitForAttribute**